### PR TITLE
More general implementations of Connection trait for r2d2 PooledConne…

### LINKED ISSUE #2027: Implement `Connection` for all `r2d2::PooledConnection<T> where T: r2d2::ManageConnection`
Currently `Connection` is implemented for `r2d2::PooledConnection<ConnectionManager<C>> where C: Connection`.

However, sometimes you might want to use a custom connection manager. In this case there is no way to implement the `Connection` trait for `r2d2::PooledConnection<MyConnectionManager>` because of the coherence rules.

Instead, diesel could provide a broader implementation for `r2d2::PooledConnection<T> where T: r2d2::ManageConnection, T::Connection: Connection`.

### PR DESCRIPTION
**Pull request by Diggsey**
…ction

Fixes #2027

### CODE REVIEW

**Comment by kiljacken** [comment]
The change could probably be made less noisy by using `M` (for manager), or even just staying with `C` as the type parameter instead of `T`, making the actual changes more visible.

**Comment by Diggsey** [comment]
Updated

**Comment by sgrif** [comment]
Seems reasonable. Thanks!

**Comment by sgrif** [comment]
Can you add a changelog entry for this?

**Comment by Diggsey** [comment]
Done.

File: diesel/src/r2d2.rs [hunk 1]
```diff
@@ -97,22 +97,24 @@ where
...
     }
 }
 
-impl<C> Connection for PooledConnection<ConnectionManager<C>>
+impl<M> Connection for PooledConnection<M>
 where
-    C: Connection<TransactionManager = AnsiTransactionManager> + Send + 'static,
-    C::Backend: UsesAnsiSavepointSyntax,
+    M: ManageConnection,
+    M::Connection: Connection<TransactionManager = AnsiTransactionManager> + Send + 'static,
```
Comment (weiznich) [inline diesel/src/r2d2.rs]: Is there any reason why we need `TransactionManager = AnsiTransactionManager` here? That will prevent connection implementation that use another TransactionManager from using this impl. (Yes I know that the original version had the same restrictions.…)

Comment (sgrif) [reply diesel/src/r2d2.rs]: Yes. If we remove it we end up with:

```
C: Connection,
C::TransactionManager: TransactionManager<Self>,
```

but the definition of `TransactionManger<C>` requires `C: Connection`, so we've got an infinite loop. https://github.com/rust-lang/rfcs/blob/master/text/1647-allow-self-in-where-clauses.md specifies behavior in this case, but that portion of the RFC has not been implemented

**Comment by Diggsey** [comment]
I'm assuming these test failures are spurious and I don't need to do anything more here?

**Pull request merged**